refacto (GPX) : Improvements and add example

gchoqueux opened the pull request:
- Use THREE.MeshLine to fix linewidth
Due to limitations in the ANGLE layer, with the WebGL renderer on Windows platforms linewidth will always be 1 regardless of the set value. 

- add pins for way points 
- fix no elevation Gpx 
- Add gpx example

![capture](https://user-images.githubusercontent.com/11291849/28825518-d6dc6b1a-76c6-11e7-8488-abb12952287e.PNG)

gchoqueux commented:
rebased

peppsac commented on src/Renderer/Lines.js:
```diff
@@ -144,11 +144,13 @@ ItownsLine.prototype.process = function process() {
 
     // les 6 points
+
+    // var indices  =  new Int16Array([0,1,2,3,4,5]);
```
Remove dead code

peppsac commented on src/Renderer/Shader/LineVS.glsl:
```diff
@@ -29,16 +6,18 @@ attribute float side;
-uniform float	MITER_LIMIT;	// 1.0: always miter, -1.0: never miter, 0.75: default
-uniform vec2	WIN_SCALE;	// the size of the viewport in pixels
+uniform float THICKNESS;  // the thickness of the line in pixels
+uniform bool MITER_LIMIT;    // 1.0: always miter, -1.0: never miter, 0.75: default
```
`bool` type conflicts with comment.

peppsac commented on src/Renderer/LineMaterial.js:
```diff
@@ -0,0 +1,44 @@
+
+    this.uniforms.THICKNESS = { value: options.linewidth };
+    this.uniforms.MITER_LIMIT = { value: true };
+    this.uniforms.WIN_SCALE = { value: new THREE.Vector2(window.innerWidth, window.innerHeight) };
```
This code is wrong (window size might be different of itowns container size).
And the container size can change so this uniforms probably needs to be updated before rendering.
(this is'nt something you'll want to do in this PR though)

peppsac replied:
Actually simply doing something similar to:
```
this.uniforms.WIN_SCALE = { value: c3DEngine().viewerDiv.size() }
```
should be enough (ie: pointing the uniforms to an Vector2 object stored elsewhere and kept up to date)

peppsac reviewed (changes requested):
Only a few minor comments to address

peppsac commented on src/Renderer/LineMaterial.js:
```diff
@@ -0,0 +1,44 @@
+
+    this.wireframe = false;
+
+    const texture = options.texture ? Fetcher.texture.load(options.texture) : undefined;
```
`Fetcher.texture(options.texture)` ?

peppsac replied:
You missed this comment.
You probably want to replace this line by:
```js
const texture = options.texture ? Fetcher.texture(options.texture).texture : undefined;
```

gchoqueux replied:
ok, I fix

peppsac reviewed (changes requested):
This PR breaks GPX rendering.
On master `itowns.viewer.loadGPX(...)` works, but the gpx isn't visible in this PR.

gchoqueux commented:
rebased

peppsac commented on src/Core/Commander/Providers/GpxUtils.js:
```diff
@@ -99,6 +101,13 @@ function _gpxToMesh(gpxXML) {
     }
 
+    gpxMesh.matrixWorld.elements = new Float64Array(16);
+    gpxMesh.matrix.elements = new Float64Array(16);
```
Add a comment explaining why you're overriding `Float32Array` with `64` ones.

peppsac commented on src/Renderer/PointMaterial.js:
```diff
@@ -13,18 +14,19 @@ const ItownsPointMaterial = function ItownsPointMaterial(options) {
     this.fragmentShader = PointFS;
 
-    var texture = new THREE.TextureLoader().load(options.texture);
+    const texture = options.texture ? Fetcher.texture(options.texture) : undefined;
```
Still not correct.
From my original comment :
```js
const texture = options.texture ? Fetcher.texture(options.texture).texture : undefined;
//                                                                ^^^^^^^^
```

peppsac commented on src/Renderer/c3DEngine.js:
```diff
@@ -32,6 +32,7 @@ function c3DEngine(scene, positionCamera, viewerDiv, debugMode, gLDebug) {
     this.scene3D.sortObjects = false;
     this.width = this.debug ? viewerDiv.clientWidth * 0.5 : viewerDiv.clientWidth;
     this.height = viewerDiv.clientHeight;
+    this.viewerDiv.size = new THREE.Vector2(this.width, this.height);
```
Don't modify `viewerDiv` please (this isn't what my original comment suggested).

peppsac replied:
Instead replace `c3DEngine`'s widhth/height property with a size and use it in like this:
```js
this.uniforms.WIN_SCALE = { value: c3DEngine().size };
```

peppsac reviewed (changes requested):
Thanks for the PR.
Do you have any idea why a .gpx line is only visible if you zoom close to the ground ? (level 13 tiles)

autra commented on examples/gpx/ULTRA2009.gpx:
```diff
@@ -0,0 +1,10683 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Why not using https://github.com/iTowns/iTowns2-sample-data ? This repo was meant for that right?

autra reviewed (dismissed):
Is this PR still relevant? (nearly all the affected files are gone) If not I guess we can close it

autra commented:
For reference, https://github.com/iTowns/itowns2/pull/345 (this PR obsoletes this the most).

gchoqueux commented:
Line's thickness doesn't work on windows

peppsac commented:
> Line's thickness doesn't work on windows

What about trying to get this fixed upstream (`three.js`) ?

gchoqueux commented:
Updated 

Reviewer?

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -84,6 +148,14 @@ function _gpxToMesh(gpxXML) {
     }
 
+    gpxMesh.matrixWorld.elements = new Float64Array(16);
+    gpxMesh.matrix.elements = new Float64Array(16);
```
You don't need to replace `.elements` anymore.

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -84,6 +148,14 @@ function _gpxToMesh(gpxXML) {
+    gpxMesh.matrixWorld.elements = new Float64Array(16);
+    gpxMesh.matrix.elements = new Float64Array(16);
+    gpxMesh.position.copy(gpxXML.center);
+    gpxMesh.updateMatrix();
```
Not needed (called by `updateMatrixWorld` if `matrixAutoUpdate` is `true` which is the default value)

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -84,6 +148,14 @@ function _gpxToMesh(gpxXML) {
+    gpxMesh.updateMatrix();
+    gpxMesh.updateMatrixWorld(true);
+    gpxMesh.matrixAutoUpdate = false;
+    gpxMesh.matrixWorldNeedsUpdate = false;
```
Not needed either (set by `updateMatrixWorld`)

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -24,40 +24,104 @@ function _gpxPtToCartesian(pt) {
         for (var i = 0; i < wayPts.length; i++) {
-            const pos = _gpxPtToCartesian(wayPts[i]);
-            pos.toArray(positions, 3 * i);
+            if (gpxXML.center === undefined) {
```
You could move this out of the `for` loop to improve readability.

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -24,40 +24,104 @@ function _gpxPtToCartesian(pt) {
+            const mesh = new THREE.Mesh(geometryPoint, materialPoint);
+            mesh.position.copy(position);
+            mesh.lookAt(gpxXML.center.clone().negate());
+            mesh.updateMatrixWorld();
```
You can save this one as well because at some point you're going to call `updateMatrixWorld`on the `points` which will then call `gpxToWayPointsMesh` recursively.

peppsac commented on examples/gpx.js:
```diff
@@ -0,0 +1,28 @@
+/* global itowns, document, renderer */
+// # Simple Globe viewer
```
Fix comment please

peppsac commented on examples/gpx.js:
```diff
@@ -0,0 +1,28 @@
+promises.push(itowns.Fetcher.json('./layers/JSONLayers/IGN_MNT_HIGHRES.json').then(addLayerCb));
+
+exports.view = globeView;
+exports.initialPosition = positionOnGlobe;
```
Unused?

peppsac commented on examples/gpx.html:
```diff
@@ -0,0 +1,71 @@
+        <script src="gpx.js"></script>
+        <script type="text/javascript">
+            /* global itowns, document, GuiTools, globeView, promises */
+            var menuGlobe = new GuiTools('menuDiv');
```
I don't think the `GuiTools` are useful in this example so I'd remove them.

This will help shortening your example to the bare minimum needed to display your gpx file.

peppsac reviewed (changes requested):
Thanks.

I left a first round of comments, mainly small style issues.

You should include a comment in the source code explaining why you're not using Points/Line in the simplest way.

peppsac commented:
Btw, I know it's out the of scope of your PR but the GPX support should support a different destination CRS than `EPSG:4978`.

Also could you add a screenshot of what GPX rendering looks like with your changes?

gchoqueux commented:
i'll add CRS support

peppsac commented on examples/gpx.html:
```diff
@@ -0,0 +1,65 @@
+        <div id="viewerDiv"></div>
+        <script src="GUI/GuiTools.js"></script>
+        <script src="../dist/itowns.js"></script>
+        <script src="../dist/debug.js"></script>
```
You can remove this line.

peppsac commented on examples/gpx.html:
```diff
@@ -0,0 +1,65 @@
+        <meta charset="UTF-8">
+
+        <meta name="viewport" content="width=device-width, initial-scale=1.0">
+        <script src="GUI/dat.gui/dat.gui.min.js"></script>
```
You can remove this line

peppsac commented on examples/gpx.html:
```diff
@@ -0,0 +1,65 @@
+                padding: 0;
+            }
+
+            #menuDiv {
```
You can remove all the `menuDiv` stuff here

peppsac commented on examples/gpx.html:
```diff
@@ -0,0 +1,65 @@
+            /* global itowns, document, GuiTools, globeView, promises */
+            // Listen for globe full initialisation event
+            globeView.addEventListener(itowns.GLOBE_VIEW_EVENTS.GLOBE_INITIALIZED, function () {
+                // eslint-disable-next-line no-console
```
`html` files are not linter so this is not needed.

peppsac commented on examples/gpx.html:
```diff
@@ -0,0 +1,65 @@
+            globeView.addEventListener(itowns.GLOBE_VIEW_EVENTS.GLOBE_INITIALIZED, function () {
+                // eslint-disable-next-line no-console
+                console.info('Globe initialized');
+                itowns.loadGpx('https://raw.githubusercontent.com/iTowns/iTowns2-sample-data/master/ULTRA2009.gpx', globeView.referenceCrs).then((gpx) => {
```
You can't use arrow functions here (must only use ES5 features)

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +16,154 @@ function _gGpxToWTrackPointsArray(gpxXML) {
     }
 
+    gpxMesh.position.copy(gpxXML.center);
+    gpxMesh.updateMatrix();
```
Call to `updateMatrix` isn't necessary.

gchoqueux commented:
updated

peppsac commented on examples/gpx.html:
```diff
@@ -0,0 +1,54 @@
+                padding: 0;
+            }
+
+            @media (max-width: 600px) {
```
You forgot to remove this as well

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +16,153 @@ function _gGpxToWTrackPointsArray(gpxXML) {
 
-export default function loadGpx(urlFile, networkOptions) {
-    return Fetcher.xml(urlFile, networkOptions).then(gpxXML => _gpxToMesh(gpxXML));
+export default function loadGpx(urlFile, crs, networkOptions) {
```
Don't you think you should include `networkOptions` in a `options` object that would also allow to configure a few things:
- use 'pin' or not
- line size (because with your change the user cannot simply set : `line.material.linewidth`)

peppsac replied:
Last but not least, but while reworking this file, maybe you could export it as:
```js
export default {
   load(urlFile, crs, options) {
   ... 
   }
};
```
This way the user would do:
```js
itowns.GpxUtils.load(...);
```

gchoqueux committed:
fix elevation problem

peppsac commented:
Another suggestion for the linewidth issue would be to use: https://github.com/spite/THREE.MeshLine/

gchoqueux commented:
@peppsac updated
Thanks, I replace by MeshLine with depth log path for shader

gchoqueux committed:
refacto (gpx) : change default export loadGpx by gpxUtils

gpxUtils.load(url) to load gpx

gchoqueux commented:
updated
add options and doc

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
-function _gpxToWTrackPointsMesh(gpxXML) {
+function updatePath(renderer, scene, camera) {
+    const distance = getDistance(this, camera);
+    this.material.depthTest = distance < this.geometry.boundingSphere.radius * 2;
```
Why do you need to disable depth test?
Is there a valid reason or is this a workaround for a bug in iTowns?

gchoqueux replied:
There's visibility problem with DTM for long distances

![sans](https://user-images.githubusercontent.com/11291849/29113217-bffafe72-7cf0-11e7-8ebe-cd0178e23c55.png)

peppsac replied:
Yep.
But my question was: did you look for the root cause (for instance playing with the near/far values of our camera) before settling on this fix?

peppsac replied:
Or is this because the elevation in the gpx and in the DEM doesn't match?

gchoqueux replied:
yes, the reason the gpx and in the DEM doesn't match
if gpx matchs with DEM, there's depth problem

peppsac replied:
Still looks like a hack to me but we can live with this for a while (#449 could help here).

peppsac commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "scripts": {
     "lint": "eslint \"src/**/*.js\" \"test/**/*.js\" \"examples/**/*.js\"",
-    "doc": "jsdoc src/Core/View.js src/Core/Prefab/GlobeView.js src/Core/Layer/Layer.js src/Renderer/ColorLayersOrdering.js src/Renderer/ThreeExtended/GlobeControls.js src/Core/Geographic/Coordinates.js",
+    "doc": "jsdoc src/Core/View.js src/Core/Prefab/GlobeView.js src/Core/Layer/Layer.js src/Renderer/ColorLayersOrdering.js src/Renderer/ThreeExtended/GlobeControls.js src/Core/Geographic/Coordinates.js src/Core/Geographic/Coordinates.js src/Core/Scheduler/Providers/GpxUtils.js",
```
`src/Core/Geographic/Coordinates.js` is now listed twice.

peppsac commented on package.json:
```diff
@@ -40,7 +40,8 @@
     "three": "^0.86.0",
-    "whatwg-fetch": "^2.0.2"
+    "whatwg-fetch": "^2.0.2",
+    "three.meshline": "^1.0.3"
```
Did you add this one manually (because it's not properly sorted)?

peppsac replied:
Also, you didn't update `package-lock.json` (that requires a recent enough version of npm)

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -5,8 +5,12 @@
 import Fetcher from './Fetcher';
 import Coordinates from '../../Geographic/Coordinates';
+import Capabilities from '../../System/Capabilities';
+// eslint-disable-next-line
```
Why do you need to disable eslint here?

gchoqueux replied:
I have to write $3dTiles_Provider to import patchMaterial otherwise there's a error
and i not use 3dTiles_Provider so i disable eslint

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -5,8 +5,12 @@
 import Coordinates from '../../Geographic/Coordinates';
+import Capabilities from '../../System/Capabilities';
+// eslint-disable-next-line
+import $3dTiles_Provider, { patchMaterial } from './3dTiles_Provider';
```
You should only import `patchMaterial`:
```js
import { patchMaterial } from './3dTiles_Provider'; 
```

gchoqueux replied:
I have to write $3dTiles_Provider to import patchMaterial otherwise there's a error

peppsac commented on src/Core/Scheduler/Providers/3dTiles_Provider.js:
```diff
@@ -99,7 +99,7 @@ function getBox(volume, inverseTileTransform) {
 const rePosition = new RegExp('gl_Position.*(?![^]*gl_Position)');
 const reMain = new RegExp('[^\\w]*main[^\\w]*(void)?[^\\w]*{');
-function patchMaterial(material) {
+export function patchMaterial(material) {
```
Maybe you could have moved this function to its own utility file.
At the very least change its name to really reflect what it's doing (`patchMaterialForLogDepthSupport`?)

gchoqueux replied:
in a new PR

peppsac replied:
Nope: this PR exports a previously private function so you must either split it to a new file or nename it here.

gchoqueux replied:
ok to rename

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
+const point = new THREE.Vector3();
+
+function getDistance(object, camera) {
+    point.setFromMatrixPosition(object.matrixWorld).add(object.geometry.boundingSphere.center);
```
You don't account for scaling and rotation here.
```js
object.geometry.boundingSphere.center.clone().applyMatrix4(object.matrixWorld);
```
would be correct I think.

gchoqueux replied:
yes

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
+
+function getDistance(object, camera) {
+    point.setFromMatrixPosition(object.matrixWorld).add(object.geometry.boundingSphere.center);
+    return camera.position.distanceTo(point);
```
This ignores camera's transformation as well (but this isn't the only place in itowns that assumes that the camera has no parent)

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
+    const distance = getDistance(this, camera);
+    this.material.depthTest = distance < this.geometry.boundingSphere.radius * 2;
+    const size = renderer.getSize();
+    this.material.uniforms.resolution.value.set(size.width, size.height);
```
We should probably provide an easy way to register things that needs to be updated on resolution change (material like this one, render targets, etc).
(not for this PR though)

gchoqueux replied:
yes

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
+    this.material.depthTest = distance < this.geometry.boundingSphere.radius * 2;
+    const size = renderer.getSize();
+    this.material.uniforms.resolution.value.set(size.width, size.height);
+    this.material.needsUpdate = true;
```
I don't think you need this. Did you test without it?

gchoqueux replied:
Forget about displacement map

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
-        for (var i = 0; i < trackPts.length; i++) {
-            const pos = _gpxPtToCartesian(trackPts[i]);
-            pos.toArray(positions, 3 * i);
+        if (gpxXML.center === undefined) {
```
`gpxXML.center = gpxXML.center || _gpxPtToCartesian(trackPts[0], crs);` ?

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
-            pos.toArray(positions, 3 * i);
+        const points = new THREE.Group();
+
+        if (gpxXML.center === undefined) {
```
`gpxXML.center = gpxXML.center || _gpxPtToCartesian(trackPts[0], crs);` ?

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
 }
 
-function _gpxToMesh(gpxXML) {
+function _gpxToMesh(gpxXML, crs, options) {
```
`, options = {}) {`

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +20,173 @@ function _gGpxToWTrackPointsArray(gpxXML) {
+     * @return {THREE.Mesh} Three.js Mesh see {@link https://threejs.org/docs/#api/objects/Mesh}
+     * @example
+     * // How add gpx object
+     * itowns.GpxUtils.load(url).then((gpx) => {
```
Missing `crs` arg.

peppsac commented on src/Core/View.js:
```diff
@@ -28,7 +28,7 @@ import Scheduler from './Scheduler/Scheduler';
  * @example
  * // How add gpx object
- * itowns.loadGpx(url).then((gpx) => {
+ * itowns.GpxUtils.load(url).then((gpx) => {
```
Missing `crs`

gchoqueux committed:
feature (gpx) : add support crs in gpx

gchoqueux committed:
refacto (gpx) : add options and documention

options.enablePin=true draw pin for way points
options.networkOptions Options for fetching resources over network
options.lineWidth = 12 set lineWidth to track line

gchoqueux committed:
feature (GPX) : scale path and point GPX to correct display

gchoqueux commented:
updated

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +19,172 @@ function _gGpxToWTrackPointsArray(gpxXML) {
         return undefined;
     }
 
+    options = options || {};
```
You can remove this, if you make sure that `options` is always defined.
And the simplest way to do this is to fix the only entry point for this module:
```js
load(urlFile, crs, options = {}) {
...
```

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +19,172 @@ function _gGpxToWTrackPointsArray(gpxXML) {
+     *
+     */
+    load(urlFile, crs, options) {
+        options = options || {};
```
As stated earlier: move this to function signature.

peppsac commented on src/Core/Scheduler/Providers/GpxUtils.js:
```diff
@@ -16,78 +19,172 @@ function _gGpxToWTrackPointsArray(gpxXML) {
+    load(urlFile, crs, options) {
+        options = options || {};
+        options.crs = crs;
+        return Fetcher.xml(urlFile, options ? options.networkOptions : {}).then(gpxXML => _gpxToMesh(gpxXML, options));
```
Then you can directly use: `.xml(urlFile, options.networkOptions)`

peppsac reviewed (approved):
I've left a few nits to address before merging but otherwise it looks good.
Thanks!
